File: scripts/generate_fireredvad_fixture.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = ["onnxruntime", "soundfile", "numpy"]
# ///
"""Golden per-frame probabilities for the FireRedVAD provider.

Reference: the fireredvad_stream_vad_e2e.onnx model run with Python
onnxruntime over the full utterance (zero initial caches). The model's own
parity with the FireRedVAD Python pipeline is established in that repo's
tests; this fixture pins the TypeScript glue (framing, cache threading).
"""

import json
import os

import numpy as np
import onnxruntime
import soundfile as sf

HERE = os.path.dirname(os.path.abspath(__file__))


def main():
    wav, sr = sf.read(
        os.path.join(HERE, "..", "tests", "assets", "hello_en.wav"), dtype="int16")
    assert sr == 16000
    pcm = (wav / 32768.0).astype(np.float32)[None, :]

    sess = onnxruntime.InferenceSession(
        os.path.join(HERE, "..", "models", "fireredvad_stream_vad_e2e.onnx"))
    caches = np.zeros((1, 1024, 19), dtype=np.float32)
    (probs,) = sess.run(["probs"], {"pcm": pcm, "caches_packed": caches})

    out = os.path.join(HERE, "..", "tests", "fixtures", "fireredvad.json")
    os.makedirs(os.path.dirname(out), exist_ok=True)
    with open(out, "w") as f:
        json.dump({"wav": "hello_en.wav", "probs": probs[0, :, 0].tolist()}, f)
    print(f"wrote {out} ({probs.shape[1]} frames)")


if __name__ == "__main__":
    main()

File: scripts/generate_fireredvad_fixture.ts
// Golden per-frame probabilities for the FireRedVAD provider.
//
// Reference: the fireredvad_stream_vad_e2e.onnx model run with
// onnxruntime-node (the same native ONNX Runtime core as Python
// onnxruntime) over the full utterance with zero initial caches. The
// model's own parity with the FireRedVAD Python pipeline is established in
// that repo's tests; this fixture pins the TypeScript glue (framing, cache
// threading) against an independent runtime.
//
// Run: npm run fixtures

import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import * as ort from "onnxruntime-node";

import { readWav16kMono } from "../tests/wav.ts";

const HERE = path.dirname(fileURLToPath(import.meta.url));

const wavBuf = readFileSync(path.join(HERE, "..", "tests", "assets", "hello_en.wav"));
const pcm = readWav16kMono(
  wavBuf.buffer.slice(wavBuf.byteOffset, wavBuf.byteOffset + wavBuf.byteLength),
);

const session = await ort.InferenceSession.create(
  path.join(HERE, "..", "models", "fireredvad_stream_vad_e2e.onnx"),
);
const outputs = await session.run({
  pcm: new ort.Tensor("float32", pcm, [1, pcm.length]),
  caches_packed: new ort.Tensor("float32", new Float32Array(1024 * 19), [1, 1024, 19]),
});
const probs = outputs.probs;
if (probs === undefined) throw new Error("model did not return probs");

const out = path.join(HERE, "..", "tests", "fixtures", "fireredvad.json");
writeFileSync(
  out,
  JSON.stringify({ wav: "hello_en.wav", probs: Array.from(probs.data as Float32Array) }) + "\n",
);
console.log(`wrote ${out} (${String(probs.dims[1])} frames)`);
